replication-api-impl/src/main/java/com/connexta/replication/api/impl/worker/WorkerThreadPool.java

```diff
@@ -0,0 +1,160 @@
+
+  private final NodeAdapters nodeAdapters;
+
+  private final Site localSite;
+
+  private final Site destinationSite;
+
+  private final Queue queue;
+
+  private AtomicInteger size = new AtomicInteger(0);
+
+  /**
+   * Creates a new worker thread pool for the given site and spins up a number of workers equal to
+   * the size passed in.
+   *
+   * @param queue queue which workers will poll tasks from
+   * @param size the initialize size of the pool
+   * @param localSite the local site
+   * @param remoteSite the remote site corresponding to the queue
+   * @param nodeAdapters factory for creating node adapters
+   */
+  public WorkerThreadPool(Queue queue, int size,
+      Site localSite,
+      Site remoteSite,
+      NodeAdapters nodeAdapters) {
+    this.queue = queue;
+    this.localSite = localSite;
+    this.destinationSite = remoteSite;
+    this.nodeAdapters = nodeAdapters;
+    setSize(size);
+  }
+
+  /**
+   * Sets the size of this thread pool. If the current size exceeds the new size, then workers
+   * workers will be interrupted until the new size is achieved. Workers that are currently not
+   * processing will be prioritized for cleanup before workers that are processing. If all workers
+   * are processing or there is a remaining amount that needs to be interrupted, then they will be
+   * stopped in the order they are encountered until the amount of workers un-interrupted is equal
+   * to the new size.
+   *
+   * If the new size is greater than the current size, then newSize-currentSize workers will be spun
+   * up.
+   *
+   * @param size the size of this thread pool
+   * @throws IllegalStateException if this thread pool has been shutdown
+   */
+  public void setSize(int size) {
+    synchronized (lock) {
+      if (shutdown) {
+        throw new IllegalStateException("This thread is shutdown");
+      }
+    }
+
+    final int currentSize = this.getSize();
+    if (size == currentSize) {
+      return;
+    }
+
+    if (size > currentSize) {
+      final int delta = size - currentSize;
+      spinUpWorkers(delta);
+      return;
+    }
+
+    final int delta = currentSize - size;
+    spinDownWorkers(delta);
+  }
+
+  private void spinDownWorkers(int target) {
+    int cancelled = 0;
+    // try to cancel non processing workers first
+    for (Worker worker : workers) {
+      if (!worker.isProcessing()) {
+        worker.cancel();
+        cancelled++;
+      }
+
+      if (cancelled == target) {
+        break;
+      }
+    }
+
+    // begin cancelling processing workers in order until target is reached
+    if (cancelled < target) {
+      for (Worker worker : workers) {
+        worker.cancel();
+        cancelled++;
+
+        if (cancelled == target) {
+          break;
+        }
+      }
+    }
+  }
+
+  private void spinUpWorkers(int num) {
+    for (int i = 0; i < num; i++) {
+      NodeAdapter localAdapter = nodeAdapters.factoryFor(localSite.getType())
```

Comment:
✏️ Since when we move to external microservices, the worker will be responsible for getting their own adapter(s). I would suggest we do not do this here and leave it up to the constructor of the Worker to do that.